packages/client/src/features/smart-connector/smart-connector.ts

```diff
@@ -0,0 +1,524 @@
+const GROUP_CLASS = 'smart-connector-group';
+const TOOL_BUTTON_CLASS = 'smart-connector-button';
+const COLLAPSABLE_CLASS = 'collapsable-group';
+
+@injectable()
+export class SmartConnector extends AbstractUIExtension implements IActionHandler, IDiagramStartup {
+
+    static readonly ID = 'smart-connector';
+
+    protected selectedElementId: string;
+    protected smartConnectorItems: SmartConnectorGroupItem[];
+    protected smartConnectorContainer: HTMLElement;
+    protected expandButton: HTMLElement;
+    protected currentZoom: number;
+
+    protected smartConnectorGroups: Record<string, HTMLElement> = {};
+    protected groupIsCollapsed: Record<string, boolean> = {};
+    protected groupIsTop: Record<string, boolean> = {};
+    protected searchFields: Record<string, HTMLInputElement> = {};
+
+    protected previousElementKeyCode: KeyCode = 'ArrowUp';
+    protected nextElementKeyCode: KeyCode = 'ArrowDown';
+
+    // Sets the position of the expand button
+    protected expandButtonPosition = SmartConnectorPosition.Top;
+
+    @inject(GLSPActionDispatcher)
+    protected actionDispatcher: GLSPActionDispatcher;
+
+    @inject(EditorContextService)
+    protected editorContext: EditorContextService;
+
+    @inject(FocusTracker)
+    protected focusTracker: FocusTracker;
+
+    override id(): string {
+        return SmartConnector.ID;
+    }
+    override containerClass(): string {
+        return SmartConnector.ID;
+    }
+
+    protected override async onBeforeShow(_containerElement: HTMLElement, root: Readonly<GModelRoot>): Promise<void> {
+        const viewportResult: ViewportResult = await this.actionDispatcher.request(GetViewportAction.create());
+        await this.requestAvailableOptions(root.index.getById(this.selectedElementId));
+        this.currentZoom = viewportResult.viewport.zoom;
+        const nodeFromDom = document.getElementById(`${this.options.baseDiv}_${this.selectedElementId}`) as any as SVGGElement;
+        const nodeBoundsFromDom = nodeFromDom.getBoundingClientRect();
+        this.setMainPosition(viewportResult, nodeBoundsFromDom);
+        // set position of expand button
+        this.setPosition(this.expandButton, this.expandButtonPosition, nodeBoundsFromDom);
+        // set position of container(s)
+        const sameSide = this.smartConnectorItems.every(e => e.position === this.smartConnectorItems[0].position);
+        if (sameSide)
+            {this.setPosition(this.smartConnectorContainer, this.smartConnectorItems[0].position, nodeBoundsFromDom);}
+        else {
+            for (let i = 0; i < this.smartConnectorContainer.childElementCount; i++) {
+                this.setPosition(this.smartConnectorContainer.children[i] as HTMLElement, this.smartConnectorItems[i].position,
+                    nodeBoundsFromDom, true);
+            }
+        }
+        this.hideSmartConnector();
+    }
+
+    protected async requestAvailableOptions(contextElement?: GModelElement): Promise<void> {
```

Comment:
This method does more than one thing:
1) initializes the Options
2) creates the body

for this reason, i would say, move `createBody`to `onBeforeShow` and rename this method to `initAvailableOptions`